# Khám phá dữ liệu và Tiền xử lý dữ liệu

## Các thư viện cần thiết

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import re

## **_01. Exploring Data:_**

In [21]:
pd.set_option('display.max_columns', None)
manga_df = pd.read_csv('./Data/manga.csv')
manga_df

manga_id                                              title  \
0             2                                            Berserk   
1            13                                          One Piece   
2            25                                Fullmetal Alchemist   
3          1706    JoJo no Kimyou na Bouken Part 7: Steel Ball Run   
4          4632                                     Oyasumi Punpun   
...         ...                                                ...   
67268    149662                    Scary Campus College University   
67269    149663  Dorei Tensei: Sono Dorei, Saikyou no Moto Ouji...   
67270    149664                             To Have You Once Again   
67271    149666                                             兄だったモノ   
67272    149667                                    Utsukushii Kare   

              type  score  scored_by                status  volumes  chapters  \
0            manga   9.45     268737  currently_publishing      NaN       NaN   
1            manga   9.20     305917  currently_publishing      NaN       NaN   
2            manga   9.05     143879              finished     27.0     116.0   
3            manga   9.27     125482              finished     24.0      96.0   
4            manga   9.03     141146              finished     13.0     147.0   
...            ...    ...        ...                   ...      ...       ...   
67268        manga    NaN          0  currently_publishing      NaN       NaN   
67269  light_novel    NaN          0  currently_publishing      NaN       NaN   
67270       manhua    NaN          0  currently_publishing      NaN       NaN   
67271        manga    NaN          0  currently_publishing      NaN       NaN   
67272        manga    NaN          0              finished      1.0       4.0   

       start_date    end_date  members  favorites   sfw  approved  \
0      1989-08-25         NaN   551266     103820  True      True   
1      1997-07-22         NaN   501291      99526  True      True   
2      2001-07-12  2010-09-11   264571      28519  True      True   
3      2004-01-19  2011-04-19   209123      35456  True      True   
4      2007-03-15  2013-11-02   354176      43168  True      True   
...           ...         ...      ...        ...   ...       ...   
67268  2022-07-25         NaN        0          0  True     False   
67269  2020-09-30         NaN        0          0  True     False   
67270         NaN         NaN        0          0  True     False   
67271  2021-05-27         NaN        0          0  True     False   
67272         NaN         NaN        0          0  True     False   

                      created_at_before                 updated_at  \
0             2007-07-17 20:14:45+00:00  2022-06-23 08:30:44+00:00   
1             2007-07-17 20:14:45+00:00  2022-04-18 04:48:35+00:00   
2             2007-07-17 20:14:45+00:00  2022-06-13 18:46:09+00:00   
3             2007-10-07 08:14:20+00:00  2022-04-18 04:02:35+00:00   
4             2008-02-03 15:54:30+00:00  2022-04-18 04:47:05+00:00   
...                                 ...                        ...   
67268         2022-07-25 00:02:54+00:00                        NaN   
67269         2022-07-25 00:02:54+00:00                        NaN   
67270         2022-07-25 00:02:54+00:00                        NaN   
67271  2022-07-25 17:32:10.227761+00:00                        NaN   
67272  2022-07-25 17:32:10.227761+00:00                        NaN   

      real_start_date real_end_date  \
0          1989-08-25           NaN   
1          1997-07-22           NaN   
2          2001-07-12    2010-09-11   
3          2004-01-19    2011-04-19   
4          2007-03-15    2013-11-02   
...               ...           ...   
67268      2022-07-25           NaN   
67269      2020-09-30           NaN   
67270             NaN           NaN   
67271      2021-05-27           NaN   
67272             NaN           NaN   

                                                  genr

#### Dữ liệu có bao nhiêu dòng và cột ?

In [3]:
manga_df.shape

(67273, 31)

Dữ liệu có 67273 dòng và 31 cột

#### Ý nghĩa của từng dòng là gì ?

Mỗi dòng là thông tin của một manga

#### Có tồn tại dòng nào bị trùng lắp không ?

In [4]:
np.sum(manga_df.duplicated())

0

Dữ liệu không có dòng nào trùng lắp

#### Ý nghĩa của từng cột là gì ?

  | Cột               | Ý nghĩa                                                                       |
  |-------------------|-------------------------------------------------------------------------------|
  |`manga_id`         | ID của manga (mỗi manga có một ID riêng biệt)                                 |
  |`title`            | Tựa đề của manga (viết theo phiên âm tiếng Nhật hoặc tiếng Anh)               |
  |`type`             | Được phát hành theo dạng nào (truyện tranh, tiểu thuyết,...)                  |
  |`score`            | Điểm số trên trang MyAnimeList (MAL)                                          |
  |`scored_by`        | Số lượng người dùng đã đánh giá manga                                         |
  |`status`           | Tình trạng (đang phát hành, đã hoàn thành, tạm dừng,...)                      |
  |`volumes`          | Số tập                                                                        |
  |`chapters`         | Số chương                                                                     |
  |`start_date`       | Ngày bắt đầu phát hành                                                        |
  |`end_date`         | Ngày kết thúc phát hành                                                       |
  |`members`          | Số lượng người dùng có manga này trong list của họ                            |
  |`favorites`        | Số lượng người dùng yêu thích manga này                                       |
  |`sfw`              | Manga có nhãn 18+ hay không                                                   |
  |`approved`         | Manga có được phê duyệt hay chưa                                              |
  |`created_at_before`| Ngày trước khi mục của manga này được tạo                                     |
  |`updated_at`       | Ngày cập nhật gần đây nhất của manga này                                      |
  |`real_start_date`  | Ngày bắt đầu phát hành thật sự                                                |
  |`real_end_date`    | Ngày kết thúc phát hành thật sự                                               |
  |`genres`           | Thể loại                                                                      |
  |`themes`           | Chủ đề                                                                        |
  |`demographics`     | Nhân khẩu học                                                                 |
  |`authors`          | Thông tin tác giả                                                             |
  |`serializations`   | Danh sách các tạp chí nơi manga này được đăng lên                             |
  |`sypnosis`         | Mô tả                                                                         |
  |`background`       | Thông tin thêm                                                                |
  |`main_picture`     | URL hình ảnh chính của manga trên MAL                                         |
  |`url`              | URL của manga trên MAL                                                        |
  |`title_english`    | Tựa đề tiếng Anh của manga                                                    |
  |`title_japanese`   | Tựa đề tiếng Nhật của manga                                                   |
  |`title_synonyms`   | Danh sách các tựa đề đồng nghĩa                                               |
  |`jikan`            | Manga này được scrape từ Jikan API hay API chính thức của MAL                 |

#### Kiểu dữ liệu hiện tại của từng cột là gì ? Có tồn tại những cột nào có kiểu dữ liệu không phù hợp hay không ?

In [5]:
manga_df.dtypes

manga_id               int64
title                 object
type                  object
score                float64
scored_by              int64
status                object
volumes              float64
chapters             float64
start_date            object
end_date              object
members                int64
favorites              int64
sfw                     bool
approved                bool
created_at_before     object
updated_at            object
real_start_date       object
real_end_date         object
genres                object
themes                object
demographics          object
authors               object
serializations        object
synopsis              object
background            object
main_picture          object
url                   object
title_english         object
title_japanese        object
title_synonyms        object
jikan                   bool
dtype: object

#### Để hiểu rõ hơn về các cột có kiểu là object, ta sẽ phân tích ra thành từng kiểu con

In [31]:
def open_object_dtype(s):
    dtypes = set()
    dtypes.update(s.apply(type))
    return dtypes

manga_df[manga_df.dtypes.index[manga_df.dtypes == object]].apply(open_object_dtype)

title                                 {<class 'str'>}
type                                  {<class 'str'>}
status                                {<class 'str'>}
start_date           {<class 'str'>, <class 'float'>}
end_date             {<class 'str'>, <class 'float'>}
created_at_before                     {<class 'str'>}
updated_at           {<class 'str'>, <class 'float'>}
real_start_date      {<class 'str'>, <class 'float'>}
real_end_date        {<class 'str'>, <class 'float'>}
genres                                {<class 'str'>}
themes                                {<class 'str'>}
demographics                          {<class 'str'>}
authors                               {<class 'str'>}
serializations                        {<class 'str'>}
synopsis             {<class 'str'>, <class 'float'>}
background           {<class 'str'>, <class 'float'>}
main_picture         {<class 'str'>, <class 'float'>}
url                                   {<class 'str'>}
title_english        {<class

#### Các cột phù hợp và không phù hợp về kiểu dữ liệu
- manga_id -> chapters, members, favorites, sfw, appoved, url, jikan: phù hợp
- start_date, end_date, created_at_before, updated_at, real_start_date, real_end_date: chuyển về date
- genres, themes, demographics, authors, serializations, title_synonyms: chuyển về str array
- synopsis, background, main_picture, title_english, title_japanese: chuyển về str

## **_02. Pre-processing Data:_**

**Ta sẽ chuyển những cột có kiểu không phù hợp về dạng thích hợp như đã đề cập ở phần 1**

## **_03. Exploring Data (cont):_**

#### Phân bổ các cột kiểu số

#### Phân bổ các cột kiểu phân loại

#### Hiểu thêm về dữ liệu (option)